Docs misc 15 (#2110)
LinuxCNC/linuxcnc: docs/src/remap/remap.adoc

The options of the REMAP statement are separated by whitespace. The options are
keyword-value pairs and currently are:

> ``modalgroup=``_<modal group>_::

Reviewer: Not sure why, but ascidoc-py likes it only this way:
```suggestion
`modalgroup=`__<modal group>__::
```

Author: Have you tested this? On my viewer my version was fine.

Reviewer: Of course. Would be nice if _you_ do this - this would save me much time and would lead to a quicker acceptance of your PRs ;-)
You remember we realized that the previewer corresponds with asciidoctor's syntax? So that is only a partial help.

  T,S,F;; for these the modal group is fixed and any
    `modalgroup=` option is ignored.

``argspec=``_<argspec>_::
  See <<sub:argspec-parameter, description of the argspec
  parameter options>>. Optional.

``ngc=``_<ngc_basename>_::
  Basename of an O-word subroutine file name. Do not specify an
  .ngc extension. Searched for in the directories specified in
  the directory specified in `[DISPLAY]PROGRAM_PREFIX`, then in
  `[RS274NGC]SUBROUTINE_PATH`. Mutually exclusive with
  `python=`. It is an error to omit both `ngc=` and  `python=`.

``python=``_<Python function name>_::
  Instead of calling an ngc O-word procedure call a Python
  function. The function is expected to be defined in the
  `module_basename.oword`
  module. Mutually exclusive with `ngc=`.

``prolog=``_<Python function name>_::
  Before executing an ngc procedure, call this Python function.
  The function is expected to be defined in the
  `module_basename.remap`
  module. Optional.

> ``epilog=``_<Python function name>_::

Reviewer: same here

Note that while many combinations of argspec options are possible, not
all of them make sense. The following combinations are useful idioms:

``argspec=``_<words>_ ``ngc=``_<procname>_ ``modalgroup=``_<group>_::
  The recommended way to call an NGC procedure with a standard argspec parameter
  conversion. Used if argspec is good enough. Note it's not good
  enough for remapping the Tx and M6/M61 tool change codes.

> ``prolog=``_<pythonprolog>_ ``ngc=``_<procname>_ ``epilog=``_<pythonepilog>_ ``modalgroup=``_<group>_::

Reviewer: and here

  internal functions may be accessed from the prolog and epilog
  handlers. Also, a longer rope to hang yourselves.

``python=``_<pythonfunction>_ ``modalgroup=``_<group>_::
  Directly call to a Python function without any argument conversion.
  The most powerful way of remapping a code and going straight to
  Python. Use this if you don't need an NGC procedure, or NGC is
  just getting in your way.

> ``argspec=``_<words>_ ``python=``_<pythonfunction>_ ``modalgroup=``_<group>_::

Reviewer: and here

Author: should all be fine now in the latest upload

$ echo 'import remap' >toplevel.py
---------------------------------------------------------------------

> Now edit ``/home/user/``_xxx_``/``_xxx_``.ini`` and add the following:

Reviewer: and here

---------------------------------------------------------------------

> The functions 'prepare_prolog' and 'prepare_epilog' are part of the _standard glue_

Reviewer: ```suggestion
The functions _prepare_prolog_ and _prepare_epilog_ are part of the _standard glue_
```
make it future proof and consistent

Author: So was I and I prefer my version. prepare_prolog is just a reference to a function that has that name, so it is ok to quote in single quotes.  The standard glue however is more of a concept and thus not quoted but emphasized.

good enough, use the ``(abort, ``_<message>_``)`` feature to terminate the handler with an
error message. Note that you can substitute numbered, named, INI and
HAL parameters in the text like
in this example (see also `tests/interp/abort-hot-comment/test.ngc`):

[source,{ngc}]
----
o100 if [..] (some error condition)
     (abort, Bad Things! p42=#42 q=#<q> INI=#<_ini[a]x> pin=#<_hal[component.pin])
o100 endif
----

NOTE: INI and HAL variable expansion is optional and can be disabled in
the <<sub:ini:sec:rs274ngc,INI file>>

If more fine grained recovery action is needed, use the idiom
laid out in the previous example:

- Define an epilog function, even if it's just to signal an error condition,
- pass a negative value from the handler to signal the error,
- inspect the return value in the epilog function,
- take any recovery action needed,
- return the error message string from the handler,
  which will set the interpreter error message and abort the program
  (pretty much like `abort, message=`).

This error message will be displayed in the UI, and returning
INTERP_ERROR will cause this error handled like any other runtime error.

> Note that both ``(abort, ``_msg_``)`` and returning INTERP_ERROR from an

Reviewer: This works
```suggestion
Note that both `(abort,` __<msg>__ `)` and returning INTERP_ERROR from an
```

`"a message text"`::
  returning a string from a handler means 'this is an error message,

>   abort the program'. Works like ``(abort, ``_msg_``)``.

Reviewer: same here

When the changer responds by asserting "tool-prepared", it considers
the prepare phase to be completed and signals `task` to

> continue. Aagain, this 'wait' is not strictly necessary IMO.

Reviewer: ```suggestion
continue. Again, this 'wait' is not strictly necessary IMO.
```
